src/server/auth/server/api_server.go

```diff
@@ -773,41 +805,114 @@ func (a *apiServer) Authenticate(ctx context.Context, req *authclient.Authentica
-	username, err := GitHubTokenToUsername(ctx, req.GitHubToken)
-	if err != nil {
-		return nil, err
+	// verify whatever credential the user has presented, and write a new
+	// Pachyderm token for the user that their credential belongs to
+	var pachToken string
+	switch {
+	case req.GitHubToken != "":
+		// Determine caller's Pachyderm/GitHub username
+		username, err := GitHubTokenToUsername(ctx, req.GitHubToken)
+		if err != nil {
+			return nil, err
+		}
+
+		// If the cluster's enterprise token is expired, only admins may log in
+		if err := a.expiredClusterAdminCheck(username); err != nil {
+			return nil, err
+		}
+
+		// Generate a new Pachyderm token and write it
+		pachToken = uuid.NewWithoutDashes()
+		if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			tokens := a.tokens.ReadWrite(stm)
+			return tokens.PutTTL(hashToken(pachToken),
+				&authclient.TokenInfo{
+					Subject: username,
+					Source:  authclient.TokenInfo_AUTHENTICATE,
+				},
+				defaultTokenTTLSecs)
+		}); err != nil {
+			return nil, fmt.Errorf("error storing auth token for user \"%s\": %v", username, err)
+		}
+
+	case req.PachAuthenticationCode != "":
+		if _, err := col.NewSTM(ctx, a.etcdClient, func(stm col.STM) error {
+			// read short-lived authentication code (and delete it if found)
+			codes := a.authenticationCodes.ReadWrite(stm)
+			key := hashToken(req.PachAuthenticationCode)
+			var tokenInfo authclient.TokenInfo
+			if err := codes.Get(key, &tokenInfo); err != nil {
+				return err
+			}
+			codes.Delete(key)
+
+			// If the cluster's enterprise token is expired, only admins may log in
+			if err := a.expiredClusterAdminCheck(tokenInfo.Subject); err != nil {
+				return err
+			}
+
+			// write long-lived pachyderm token
+			pachToken = uuid.NewWithoutDashes()
+			return a.tokens.ReadWrite(stm).PutTTL(hashToken(pachToken), &tokenInfo,
+				defaultTokenTTLSecs)
+		}); err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, fmt.Errorf("unrecognized authentication mechanism (old pachd?)")
 	}
 
-	// If the cluster's enterprise token is expired, only admins may log in
-	state, err := a.getEnterpriseTokenState()
+	// Return new pachyderm token to caller
+	return &authclient.AuthenticateResponse{
+		PachToken: pachToken,
+	}, nil
+}
+
+func (a *apiServer) GetAuthenticationCode(ctx context.Context, req *authclient.GetAuthenticationCodeRequest) (resp *authclient.GetAuthenticationCodeResponse, retErr error) {
+	switch a.activationState() {
+	case none:
+		// PPS is authenticated by a token read from etcd. It never calls or needs
+		// to call authenticate, even while the cluster is partway through the
+		// activation process
+		return nil, authclient.ErrNotActivated
+	case partial:
+		return nil, authclient.ErrPartiallyActivated
+	}
+	// We don't want to actually log the request/response since they contain
+	// credentials.
+	defer func(start time.Time) { a.LogResp(nil, nil, retErr, time.Since(start)) }(time.Now())
```

Comment:
Might be good to move this to above the switch, since that can cause the call to error and we may want to see logs for it.